### PR TITLE
TNT-39260 [ODD] Implement artifact provider

### PR DESCRIPTION
**Pull request by XDex**
Implement artifact provider and add support for SDK Events

## Related Issues

https://jira.corp.adobe.com/browse/TNT-39260
https://jira.corp.adobe.com/browse/TNT-38935

### CODE REVIEW

File: Source/Adobe.Target.Client/Service/RuleLoader.cs [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+            HttpStatusCode.BadGateway,
+            HttpStatusCode.ServiceUnavailable,
+            HttpStatusCode.GatewayTimeout,
+        };
+
+        private readonly TargetClientConfig clientConfig;
+        private readonly ILogger? logger;
+        private readonly IOnDeviceDecisioningHandler? handler;
+        private Timer? timer;
+        private volatile OnDeviceDecisioningRuleSet? latestRules;
```
Comment (artur-ciocanu) [inline Source/Adobe.Target.Client/Service/RuleLoader.cs:53]: Shouldn't be `lastEtag` and `succeeded` be volatile as well, these are shared variables just like `latestRules`.

File: Source/Adobe.Target.Client/Service/RuleLoader.cs [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+
+            if (!this.DeserializeRules(artifactPayload, out var deserialized))
+            {
+                return;
+            }
+
+            Interlocked.Exchange(ref this.latestRules, deserialized);
+
+            if (!this.succeeded)
+            {
```
Comment (artur-ciocanu) [inline Source/Adobe.Target.Client/Service/RuleLoader.cs:82]: What would happen to `succeeded` if we load rules multiple times. As far as I can tell `succeeded` is set to `true` once and ready callback is invoked ONLY once as well.

File: Source/Adobe.Target.Client/Service/RuleLoader.cs [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+
+            var policyResult = await Policy
+                .Handle<HttpRequestException>()
+                .OrResult<IRestResponse>(r => HttpStatusCodesWorthRetrying.Contains(r.StatusCode))
+                .RetryAsync(MaxRetries)
+                .ExecuteAndCaptureAsync(() => client.ExecuteAsync(this.GetRequest()))
+                .ConfigureAwait(false);
+
+            this.ProcessResult(policyResult);
+
```
Comment (artur-ciocanu) [inline Source/Adobe.Target.Client/Service/RuleLoader.cs:113]: Please extract constants, no magic numbers

File: Source/Adobe.Target.Client/Service/TargetService.cs [hunk 4]
```diff
@@ -128,34 +130,14 @@ private void SetUrl(string locationHint)
...
             }
 
-            this.stickyBaseUrl = newUrl;
-            this.SetDeliveryApi();
-        }
-
-        private void SetDeliveryApi()
-        {
-            this.deliveryApi = new DeliveryApi(this.GetDeliveryApiConfig(this.stickyBaseUrl));
+            Interlocked.Exchange(ref this.stickyBaseUrl, newUrl);
```
Comment (artur-ciocanu) [inline Source/Adobe.Target.Client/Service/TargetService.cs:133]: As far as I understand we should atomically update the `DeliveryApi` client `baseUrl`, are we sure `Interocked.Exhange` for URL is enough and we don't need to protect the `DeliveryApi` via a `lock`?

File: Source/Adobe.Target.Client/TargetClient.cs [hunk 5]
```diff
@@ -45,6 +45,7 @@ public static TargetClient Create(TargetClientConfig clientConfig)
         public void Initialize(TargetClientConfig clientConfig)
         {
             this.targetService = new TargetService(clientConfig);
+            this.localService = new OnDeviceDecisioningService(clientConfig);
             this.defaultDecisioningMethod = clientConfig.DecisioningMethod;
             this.defaultPropertyToken = clientConfig.DefaultPropertyToken;
```
Comment (artur-ciocanu) [inline Source/Adobe.Target.Client/TargetClient.cs:50]: Please remove `Console.WriteLine`, use logging for debugging.

**Review by artur-ciocanu — CHANGES_REQUESTED** [review]
@XDex I have left a few comments, please take a look and resend.

I wanted to ask could we please separate PRs that are adjusting OpenAPI infrastructure from functionality that you add. It's a little bit scary when you 70+ files in a PR.

File: Source/Adobe.Target.Client/Service/RuleLoader.cs [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+    using RestSharp;
+
+    internal sealed class RuleLoader
+    {
+        private const string AcceptHeader = "Accept";
+        private const string EtagHeader = "ETag";
+        private const string NoneMatchHeader = "If-None-Match";
+        private const string AcceptHeaderValue = "application/json";
+        private const string MajorVersion = "1";
+        private const string ArtifactFilename = "/rules.json";
```
[the remark this answers is not in the document]
Comment (XDex) [reply Source/Adobe.Target.Client/Service/RuleLoader.cs:37]: I took it from Java SDK - should I add client config for it or just lower it to `5`?

[the remark this answers is not in the document]
Comment (XDex) [reply Source/Adobe.Target.Client/Service/RuleLoader.cs:37]: Lowered to 5 for now

File: Source/Adobe.Target.Client/Service/RuleLoader.cs [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+        }
+
+        private void SetLocalRules()
+        {
+            var artifactPayload = this.clientConfig.OnDeviceArtifactPayload;
+            if (string.IsNullOrEmpty(artifactPayload))
+            {
+                return;
+            }
+
```
[the remark this answers is not in the document]
Comment (XDex) [reply Source/Adobe.Target.Client/Service/RuleLoader.cs:73]: Yes, it's the recommended approach when returning multiple values and is perfect for this use-case where we can both validate deserialization and get the deserialized value in one statement: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/out-parameter-modifier#declaring-out-parameters

File: Source/Adobe.Target.Client/Service/RuleLoader.cs [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+                .ExecuteAndCaptureAsync(() => client.ExecuteAsync(this.GetRequest()))
+                .ConfigureAwait(false);
+
+            this.ProcessResult(policyResult);
+
+            this.ScheduleTimer(this.clientConfig.OnDeviceDecisioningPollingIntSecs * 1000);
+        }
+
+        private IRestRequest GetRequest()
+        {
```
[the remark this answers is not in the document]
Comment (XDex) [reply Source/Adobe.Target.Client/Service/RuleLoader.cs:118]: Fixed

File: Source/Adobe.Target.Client/Service/RuleLoader.cs (hunk 1 above)
Comment (XDex) [reply Source/Adobe.Target.Client/Service/RuleLoader.cs:53]: No, only `latestRules` will be read by external threads, `lastEtag` and `succeeded` are always read/set non-concurrently

File: Source/Adobe.Target.Client/Service/RuleLoader.cs (hunk 2 above)
Comment (XDex) [reply Source/Adobe.Target.Client/Service/RuleLoader.cs:82]: Right, it's called only once by design (same as in Java SDK), and we state this in the doc:
```
        /// <summary>
        /// This is called once when on-device execution is ready
        /// </summary>
        void OnDeviceDecisioningReady();
```

File: Source/Adobe.Target.Client/Service/RuleLoader.cs (hunk 3 above)
Comment (XDex) [reply Source/Adobe.Target.Client/Service/RuleLoader.cs:113]: Fixed

File: Source/Adobe.Target.Client/Service/TargetService.cs (hunk 4 above)
Comment (XDex) [reply Source/Adobe.Target.Client/Service/TargetService.cs:133]: Yes, just having the fields volatile and using Interlocked.Exchange is enough, see: https://stackoverflow.com/a/2193445

File: Source/Adobe.Target.Client/TargetClient.cs (hunk 5 above)
Comment (XDex) [reply Source/Adobe.Target.Client/TargetClient.cs:50]: Fixed

**Comment by XDex** [comment]
@artur-ciocanu I've pushed some fixes, please take another look when you have a moment, thanks!
Right, will make sure future openapi-related changes are in separate PRs.

**Review by artur-ciocanu — APPROVED** [review]
Looks good

**Pull request merged**